### ISSUE TITLE
This sketch runs just fine in Processing, raises nonsense error in PJS

### ISSUE THREAD

**Issue by teo1978** [processing-js/processing-js#178]
(please for god's sake configure this bug tracker to accept zip attachments, this is ridiculous)

This sketch:
matteosistisette.com/test/pjsbug/ga1.zip

uses multiple .pde files, the main sketch plus one file per class definition.
It compiles and works in Java mode (I know you cannot fully test it because it communicates with PHP scripts on a server, but you can see that it doesn't raise the nonsense error it raises in PJS).

On PJS, it issues this error:
```
Uncaught TypeError: dna.getString is not a function
$constr_4 @ VM172:136
$constr @ VM172:159
Individual @ VM172:161
rebuildCurrentPopulation @ VM172:341
loadPopulation @ VM172:319
loadCurrentPopulation @ VM172:303
setup @ VM172:266
Processing.Processing.executeSketch @ processing.js:21508
Processing.Processing @ processing.js:21535
callback @ processing.js:950
Processing.loadSketchFromSources.xhr.onreadystatechange @ processing.js:929
```

**Comment by Pomax** [comment]
it's github's bug tracker; if you want to be able add attachments like .zip files, you'll have to take that up with them. Looking at the content of the .zip file though, there's a whole lot of code that does not look revelant to the problem, creating the minimal test case can be done with a single file by removing all the functions you don't use and putting the `Dna` class definition in your main file. As is, I can't run your sketch because of at least two things: you're reaching into the Java subsystem for imports, which Processing.js does not offer, and it depends on at least one php file, which isn't in the zip.

I'd strongly recommend first reducing this to a minimal example.

**Comment by GoToLoop** [comment]
> (you're using java.util.Map. I'd recommend using Procesisng's built in HashMap, which Processing.js does support)

**_Map_** is an interface not a class. Therefore, it's irrelevant for "JS Mode" whether it's imported or not!

**Comment by Pomax** [comment]
true, but that point would have been made just as well without the quote and the exclamation points. Can I ask you to please tone down your quotes and bolds and exclamation points in these kind of comments? You made a perfectly reasonable observation, it doesn't need angry-looking shouting.

**Comment by teo1978** [comment]
As i said, missing the php is irrelevant for reproducing the issue. You should be able to reproduce the issue in pjs with the sketch as is. Though of course after fixing the issue the sketch still won't work but the errors will be very different.
Yes, a minimal example isolating the issue would be more helpful. A patch fixing the bug would be even more helpfu yet i am not able to do that. All i can provide by now is what i provided.if it helps you to fix the bug great, otherwise i'm sorry i can't help more.

The map import is irrelevant as pjs just ignores imports or at least it should, and HashMap is part of processing core and i've already tested that HashMap works in pjs

**Comment by teo1978** [comment]
And i hope it didn't sound like i'm shouting or something

**Comment by Pomax** [comment]
no, your issues are generally phrased just fine. Even if you do sound angry, it's usually because of reasonable frustration. I'll try to look at this after work, although I can't promise anything since I'm somewhat disabled at the moment due to a hand and shoulder injury

**Comment by teo1978** [comment]
Sorry about the zip thing. I thought it was configurable per project. Last time i reported a bug to Sourceforge (angrily, out of reasonable frustration, LOL) it turned out it was customizable and hence the responsibility of the project owner. Seems like this time i made the opposite wrong assumption :)

**Comment by Pomax** [comment]
yeah sourceforge did project management wildly different. I think github doesn't allow any kind of attachments, but does allow embedded data for things like images (you can even paste an image as an embed straight from the clipboard without "a file" to speak of)

**Comment by GoToLoop** [comment]
> ... without the quote...

I use quotes in order for any1 reading it to know exactly what it's being focused in the following reply.
In a discussion, it's easy to get lost when there's more than 2 folks in the subject.

> ... bolding...

I always bold (or some other available style) interfaces, classes and method names.
It got nothing to do w/ tone or scream!  @_@

> ... and the exclamation point.

Are you serious? Are you dictating the way I should type in?!

**Comment by Pomax** [comment]
I am asking you. It's up to you to decide whether you're willing to accomodate that request because someone you're talking to is explaining that your voice feels extremely inappropriate, or are going to mock that person for feeling like you're constantly yelling and thus doesn't really want to listen to what you have to say, even if you make good points? That's up to you, but at this point every time I see a post by you, I go "oh no, not again" because it's fuil of shouting and bolded text, half-tempting me to just delete your comment. Obviously I don't, but this is the effect your writing style has, and something you should probably reflect on.

**Comment by GoToLoop** [comment]
After so many wild accusations you already done to me w/o any recourse of defense, since the thread got blocked, definitely you don't treat me w/ respect.
It's sad that I even have to explain usage of bold, bullets, quotes and emojis among fellow programmers that supposed would be smart folks.
Indeed in these unwelcome circumstances, it's better off let the errors I spot follow its course.
As a last resource I might try to PM the person than being subjected to vilification and ignorance for I have been hurt too much.

**Comment by teo1978** [comment]
Somewhat more minimal code:
Replace the main sketch code with this:
```
void setup() {
  size(300,300);
  Individual individual=new Individual(1, "000000000000000000000000000000000000000", 0, 0);
}

void draw() {
 
}
```

Keep the other files as in the zip linked above.

If you comment out the println() line at the beginning of the Individual class constructor (which calls dna.getString()), calls to other methods of the Dna class will give the same error.

Putting all the code into a monolitic sketch file doesn't help either (so it's not an issue with multiple files).

**Comment by teo1978** [comment]
Ok, it was my bad. Well, almost completely.

It boils down to having two constructors in the Individual class with signatures that differ in the types but not in the number of arguments, something that is not supported in PJS.

Speaking of which...

1) Is it really impossible to support that in PJS? I do understand that in JavaScript parameters don't have types, and how PJS converts functions into javascript functions, and manages different signatures with different numbers of parameters, and how it basically ignores the parameter type declarations. But couldn't it use the parameter type declarations to create two versions of the function (just as it does with the number of parameters), and then replace the function calls with the appropriate function, given that it can know the "type" of the variable at "compile" time..... Well, maybe that would require to make the "compile" process much more complex than it currently is.

2) Even discarding that, *detecting* the presence of multiple method definitions with the same "JavaScript signature" (i.e. name and number of parameters) should definitely be possible, and hence issuing a **human-readable exception**, rather than blindly compile it into something with unpredictable behavior that will fail in a way very difficult to debug.

**Comment by Pomax** [comment]
1. unfortunately, yes. The only real solution is to build a VM inside of JS that understands typing and function overloading. In addition to being an incredibly complex task, it'd also make things run really slowly because every call would have to be type checked and disambiguated =(
2. that's not a bad idea. marking as enhancement.
